### ISSUE TITLE
SAM and the bioengine / bioimageio-colab

### ISSUE THREAD

**Issue by constantinpape** [bioimage-io/bioimageio-colab#3]
## Running SAM in the Modelzoo Universe

We have started with some efforts on integrating SAM with the bioengine / imjoy / bioimageio-colab.
I want to summarize here the overall goals, the current approaches and the steps and questions for how to achieve these goals.

### Goals

I think there are two main goals for this integration:
1. Implementing a test-run functionality for SAMs in the modelzoo website, where users can upload an example image and test how well a given model works for it (in interactive segmentation).
2. Implementing annotation tools based on SAM that can be used within other tools build with imjoy e.g. for collaborative annotation.

For now I am mostly interested in implementing goal 1, but I think goal 2 is much more interesting mid/long-term.
Ultimately it would be nice to have a set of functionality that can be used to build apps for both of these approaches.

### Current Approaches

We have two prototypes for SAM integration:
- An imjoy app that uses a model on the hypha triton server [here](https://gist.github.com/Nanguage/a09ec04b651dbcbf9692355e66240e43). (Note: that currently doesn't run, presumably because the model is not available).
  - This is the approach to implement goal 1.
- A [script](https://github.com/bioimage-io/bioimageio-colab/blob/main/bioimageio_colab/sam_server.py) that starts a server to serve a SAM model and an [app](https://github.com/bioimage-io/bioimageio-colab/blob/main/plugins/bioimageio-colab-sam.imjoy.html) that connects to it and enables point based segmentation. (The prototype works, I only tested for running server and client on the same machine).
   - This is the approach to implement goal 2. (Though it would ultimately be best to share as much functionality between these as possible).

### Next steps / Questions

- To implement the test functionality we would need a SAM model in hypha to test it. @oeway could you upload [this model](https://owncloud.gwdg.de/index.php/s/hHAAcRmOlyPWNVN) so I could test it? It contains the image encoder as torchscript and the prompt encoder and mask decoder as onnx.
- What is the best way to have a library to re-use functionality for the user-interface on the java-script side?

**Comment by oeway** [comment]
Will look more into this! Agreed to the steps forward, I would be very happy to support this.

I have already uploaded the two models you provided to the model repository, and they should be synchronized automatically to the bioengine instances:
 * https://uk1s3.embassy.ebi.ac.uk/model-repository/sam-vit_t-decoder/config.pbtxt
 * https://uk1s3.embassy.ebi.ac.uk/model-repository/sam-vit_t-encoder/config.pbtxt

For the client side, we can change the api to use the bioengine too. I will try to get back to this, or maybe @nilsmechtel can also help here too.

**Comment by constantinpape** [comment]
> I have already uploaded the two models you provided to the model repository, and they should be synchronized automatically to the bioengine instances:

Thanks! I tried to access them, but this failed:
```python
from imjoy_rpc.hypha import connect_to_server

SERVER_URL = "https://hypha.bioimage.io"


async def run():
    server = await connect_to_server(
        {"name": "test client", "server_url": SERVER_URL, "method_timeout": 100}
    )
    triton = await server.get_service("triton-client")
    config = await triton.get_config(model_name="sam-vit_t-encoder")

if __name__ == "__main__":
    import asyncio
    asyncio.run(run())
```
Fails with
```
    response = await client.get_model_config(model_name)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/lib/python3.12/site-packages/pyotritonclient/http.py", line 524, in get_model_config
    _raise_if_error(response)
  File "/opt/conda/lib/python3.12/site-packages/pyotritonclient/http.py", line 73, in _raise_if_error
    raise error
pyotritonclient.utils.InferenceServerException: Request for unknown model: 'sam-vit_t-encoder' is not found
```
Can you share a small snippet for how to correctly access the model?

> For the client side, we can change the api to use the bioengine too. I will try to get back to this, or maybe @nilsmechtel can also help here too.

That would be great! We can also set up a zoom meeting at some point to coordinate.

**Comment by oeway** [comment]
Thanks for trying, it appears that the model is failing, 
```
triton-1  | I0517 00:19:33.953968 1 libtorch.cc:1430] TRITONBACKEND_ModelInitialize: sam-vit_t-encoder (version 1)
triton-1  | W0517 00:19:33.954549 1 libtorch.cc:264] skipping model configuration auto-complete for 'sam-vit_t-encoder': not supported for pytorch backend
triton-1  | I0517 00:19:33.954834 1 libtorch.cc:293] Optimized execution is enabled for model instance 'sam-vit_t-encoder'
triton-1  | I0517 00:19:33.954851 1 libtorch.cc:311] Inference Mode is enabled for model instance 'sam-vit_t-encoder'
triton-1  | I0517 00:19:33.954859 1 libtorch.cc:406] NvFuser is not specified for model instance 'sam-vit_t-encoder'
```

I will need more investigation.

**Comment by constantinpape** [comment]
Thanks for checking again @oeway. I will check it out locally later.

**Comment by constantinpape** [comment]
Hi @oeway,
I tried it locally, but can't reproduce the error. This code works for me with the encoder:
```python
 import torch

model = torch.jit.load("test-export/sam-vit_t-encoder/1/model.pt")

input_data = torch.randn(1, 3, 1024, 1024)

model.eval()  # Set to evaluation mode
with torch.no_grad():
    output = model(input_data)

print("Run prediction ...")
print(output.shape)
```

But I went ahead and uploaded another test model, using a `vit_b` encoder [here](https://owncloud.gwdg.de/index.php/s/jmkBIsZB8ify669). Could you see if that one works in hypha / triton?
